src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -416,25 +382,41 @@ private void StartWriting(byte[] buffer, int offset, int count, AsyncProtocolReq
 
                     int chunkBytes = Math.Min(count, _sslState.MaxDataSize);
                     int encryptedBytes;
-                    SecurityStatusPal status = _sslState.EncryptData(buffer, offset, chunkBytes, ref outBuffer, out encryptedBytes);
-                    if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+                    byte[] outBuffer = s_PinnableWriteBufferCache.AllocateBuffer();
+                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
                     {
-                        // Re-handshake status is not supported.
-                        ProtocolToken message = new ProtocolToken(null, status);
-                        throw new IOException(SR.net_io_encrypt, message.GetException());
+                        PinnableBufferCacheEventSource.Log.DebugMessage3("In System.Net._SslStream.StartWriting Using Pinnable", this.GetHashCode(), count, PinnableBufferCacheEventSource.AddressOfByteArray(outBuffer));
                     }
 
-                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
+                    try
+                    {
+                        SecurityStatusPal status = _sslState.EncryptData(buffer, offset, chunkBytes, ref outBuffer, out encryptedBytes);
+
+                        if (status.ErrorCode == SecurityStatusPalErrorCode.ContinueNeeded)
+                        {
+                            chunkBytes = 0;
+                        }
+                        else if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+                        {
+                            // Re-handshake status is not supported.
+                            ProtocolToken message = new ProtocolToken(null, status);
+                            throw new IOException(SR.net_io_encrypt, message.GetException());
+                        }
+
+                    }
+                    catch
                     {
-                        PinnableBufferCacheEventSource.Log.DebugMessage3("In System.Net._SslStream.StartWriting Got Encrypted Buffer",
-                            this.GetHashCode(), encryptedBytes, PinnableBufferCacheEventSource.AddressOfByteArray(outBuffer));
+                        FreeBuffer(outBuffer);
+                        throw;
                     }
 
+
                     if (asyncRequest != null)
                     {
                         // Prepare for the next request.
                         asyncRequest.SetNextRequest(buffer, offset + chunkBytes, count - chunkBytes, s_resumeAsyncWriteCallback);
-                        Task t = _sslState.InnerStream.WriteAsync(outBuffer, 0, encryptedBytes);
+                        Task t = _sslState.InnerStream.WriteAsync(outBuffer, 0, encryptedBytes)
+                            .ContinueWith(_freeBufferAction, outBuffer);
```

Comment:
I have a few concerns here:
1. Any error from the WriteAsync call is going to get lost.
2. This is going to force the continuation to run asynchronously, even if the write complete synchronously, making it much more likely we'll hit the else path below.
3. We're paying for the continuation even if the write completes synchronously.

I suggest the following instead:
1. Remove the ContinueWith call from where it currently is.
2. For synchronous completion, add a call in the if block below to free the buffer (before doing the GetResult call).
3. For asynchronous completion, in the else block, add something like:
```C#
t = t.ContinueWith(_freeBufferAction, outBuffer, CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
```
and change _freeBufferAction to be initialized as:
```C#
_freeBufferAction = (t, buffer) =>
{
    FreeBuffer((byte[])buffer);
    t.GetAwaiter().GetResult(); // propagate any exception
};
```
That'll at least help avoid overheads in the asynchronous case.  But it's also unfortunate we're now having to create a continuation for the asynchronous case, when there's already a continuation being created as part of the TaskToApm.Begin call.  It'd be worth following up here subsequently, such that if we don't rip out the APM implementation entirely, we at least augment TaskToApm.Begin to allow it to do a bit of extra work as part of the continuation it already has, where in this case that bit of extra work would be freeing the buffer.